src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java

```diff
@@ -764,6 +817,21 @@ public String getAuthenticationGatewayUrl() {
         return "securityRealm/escapeHatch";
     }
 
+    @Override
+    public Filter createFilter(FilterConfig filterConfig) {
+        return new ChainedServletFilter(super.createFilter(filterConfig), new Filter() {
+            @Override
+            public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+                    throws IOException, ServletException {
+
+                if (OicSecurityRealm.this.handleTokenExpiration(
+                        (HttpServletRequest) request, (HttpServletResponse) response)) {
+                    chain.doFilter(request, response);
```

Comment:
It is not obvious (to me) from code inspection that this effectively runs before `BasicHeaderApiTokenAuthenticator`. The expected behavior, which I think could be asserted in `PluginTest` without new machinery beyond `TEST_USER_GROUPS_REFRESHED` etc., is that if a user’s groups change in the IdP and the previous token has expired, then a request authenticated using a Jenkins API token will in fact run under an `Authentication` reflecting that update to `authorities`. Or indeed if the user becomes completely invalid (e.g., removed from IdP realm) that a `UsernameNotFoundException` is thrown and the API token becomes unusable.

See also https://github.com/jenkinsci/oic-auth-plugin/pull/333#discussion_r1630141770: we no longer even define a custom `UserDetailsService` (what the API token processing code calls into), but on the other hand `handleTokenRefreshResponse` calls `loginAndSetUserData` which calls `fireLoggedIn` which probably accomplishes the goal. It is just not covered by the existing tests that I can see.

Reply:
From some preliminary experiments run by a colleague (@mikecirioli) it seems this does not in fact work, and could not work because the `handleTokenExpiration` method would be called in the context of an HTTP request with no servlet session (or a freshly generated empty session) and thus no `OicCredentials` and just return immediately without doing anything. In fact no approach based on keeping the refresh token (and whatever other information might be required for a refresh request) in a session attribute, or generally in memory, could possibly work; you would have to save everything required to disk, conventionally in a hidden `UserProperty`.

This PR purports to fix #100, which warns

> session cookies, API Tokens and so on may be valid forever, even if the user has long been disabled in the Identity Provider

It seems like what is fixed here is only the security of session cookies but not of API tokens; and #100 was originally filed as a security issue in response to #41, about handling API tokens.

Consider the scenario that a Jenkins user while part of some team creates an API token, perhaps for legitimate purposes, and saves it on a personal laptop. A month later they are fired and removed from the identity provider, or remain in the system but are reassigned to an unrelated team and the corresponding “group” removed from the IdP. A month after that (assume the controller has been restarted several times), _without attempting to log in to the controller again_, they check their API token (`curl`, `jenkins-cli.jar`) and find that it still works and still grants them the same permissions to reconfigure jobs or whatever they had while on the team. The only way to revoke access is for the Jenkins administrator to delete their user record from Jenkins or otherwise explicitly expire their API tokens.

Reply:
Hi, I'm still processing what you asked - but I can assure you the filter is executed correctly and right after the default security handling defined in SecurityRealm::createFilter. You may notice that the new filter is plugged right after the default filters, which also includes loading of security context from session, distributed storage or wherever it's stored. I'd like that you share how the mentioned colleague deduced that the token refresh would not happen. I do note that you mention session a lot - however I only work with Spring SecurityContext which may or may not be stored in the session.

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/5b99b75b-15bb-43e9-aa04-d0e4b02bc7e8)

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/953f6123-1327-4540-b7e5-fe053a62f693)

Reply:
I only have Keycloak that I can use for testing, and if I terminate my sessions via Keycloak console, I will get the "user refresh has expired"

Before:

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/998e5327-ea03-4db0-baf5-efb52c1a27b2)

After:

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/14e9e141-74e5-4aed-9e82-c072ab41708c)

Which yields the expected result in Jenkins:

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/22382556-4666-4c07-8258-009cf1049dbd)

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/9f082ab5-934f-4bdf-80e5-fab2ec347ba1)

Reply:
> It is not obvious (to me) from code inspection that this effectively runs before `BasicHeaderApiTokenAuthenticator`.

I have just verified this - the method runs _after_ BasicHeaderApiTokenAuthenticator, which is defined in the parent createFilters method. Should it run _before_?

Reply:
I am not sure what the last screenshot is about. The point is to use an existing API token after a user is removed from Keycloak, without having accessed the controller using a browser in the interim.

> I only work with Spring SecurityContext which may or may not be stored in the session.

Right, but so far as I can tell after the initial login flow this context would be stored in the session by `HttpSessionContextIntegrationFilter2` / `HttpSessionSecurityContextRepository` for requests coming from a web browser with a session cookie. A REST or CLI request would bypass that, so the next filter to run would be `BasicHeaderProcessor` calling `BasicHeaderApiTokenAuthenticator`. That would ask the `UserDetailsService` for an `Authentication`, which with this plugin would be the default `UserDetailsServiceProxy` with no `delegate`, always throwing `UserMayOrMayNotExistException2`; `getUserDetailsForImpersonation2` would use `ImpersonatingUserDetailsService2` which catches that and falls back to `LastGrantedAuthoritiesProperty`, picking up the groups from the last web login. That would result in a `UsernamePasswordAuthenticationToken` with a `String` principal, not a `OicCredentials` as needed to trigger the refresh flow.

Unconfirmed so far, we will try to demonstrate the issue with a test.

Reply:
No need to demonstrate, I have figured out what you talked about and will try to implement it to support the mentioned use-case.

Reply:
> if the token and/or user is no longer valid, then a request with such API Key must return 401

Right; and along the same lines, if the user is still valid but now has different group claims (at least after some expiration time has elapsed), the authentication in effect when using the API token should reflect the updated `authorities` (since these might have been used by the `AuthorizationStrategy` to make access control decisions).

Reply:
I believe I may have ran into a problem with user initiated logout. The process will invalidate the AT/RT and even if it's saved, the API key call will run into a 401 in this case. How to handle this scenario?

Reply:
Not sure if there is a solution, other than to tell users who need continued access to API tokens to not log out.